src/ln/msgs.rs

```diff
@@ -382,9 +381,95 @@ impl NetAddress {
 			&NetAddress::OnionV3 {..} => { 4 },
 		}
 	}
+
+	/// Strict byte-length of address descriptor, 1-byte type not recorded
+	/// If address is unknown, we give back a len of 0
+	fn len(&self) -> u16 {
+		match self {
+			&NetAddress::IPv4 { .. } => { 6 },
+			&NetAddress::IPv6 { .. } => { 18 },
+			&NetAddress::OnionV2 { .. } => { 12 },
+			&NetAddress::OnionV3 { .. } => { 37 },
+		}
+	}
 }
 
-#[derive(Clone)]
+impl Writeable for NetAddress {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		match self {
+			&NetAddress::IPv4 { ref addr, ref port } => {
+				1u8.write(writer)?;
+				addr.write(writer)?;
+				port.write(writer)?;
+			},
+			&NetAddress::IPv6 { ref addr, ref port } => {
+				2u8.write(writer)?;
+				addr.write(writer)?;
+				port.write(writer)?;
+			},
+			&NetAddress::OnionV2 { ref addr, ref port } => {
+				3u8.write(writer)?;
+				addr.write(writer)?;
+				port.write(writer)?;
+			},
+			&NetAddress::OnionV3 { ref ed25519_pubkey, ref checksum, ref version, ref port } => {
+				4u8.write(writer)?;
+				ed25519_pubkey.write(writer)?;
+				checksum.write(writer)?;
+				version.write(writer)?;
+				port.write(writer)?;
+			}
+		}
+		Ok(())
+	}
+}
+
+impl<R: ::std::io::Read>  Readable<R> for Result<NetAddress, u8> {
+	fn read(reader: &mut R) -> Result<Result<NetAddress, u8>, DecodeError> {
+		let byte = <u8 as Readable<R>>::read(reader)?;
+		match byte {
+			1 => {
+				let addr = Readable::read(reader)?;
```

Comment:
Note that you dont need to bind here, you can just say NetAddress::IPv4 { addr: Readable::read(reader)?, port: Readable::read(reader)? }.

Reply:
Right